Mathlib/Data/Dfinsupp/Multiset.lean

```diff
@@ -0,0 +1,116 @@
+
+This defines `Dfinsupp.toMultiset` the equivalence between `Π₀ a : α, ℕ` and `Multiset α`, along
+with `Multiset.toDfinsupp` the reverse equivalence.
+
+Note that this provides a computable alternative to `Finsupp.toMultiset`.
+-/
+
+
+variable {α : Type _} {β : α → Type _}
+
+namespace Dfinsupp
+
+/-- Non-dependent special case of `Dfinsupp.addZeroClass` to help typeclass search. -/
+instance addZeroClass' {β} [AddZeroClass β] : AddZeroClass (Π₀ _a : α, β) :=
+  @Dfinsupp.addZeroClass α (fun _ ↦ β) _
+#align dfinsupp.add_zero_class' Dfinsupp.addZeroClass'
+
+variable [DecidableEq α]
+
+/-- A computable version of `Finsupp.toMultiset`. -/
+def toMultiset : (Π₀ _a : α, ℕ) →+ Multiset α :=
+  Dfinsupp.sumAddHom fun a : α ↦ Multiset.replicateAddMonoidHom a
+#align dfinsupp.to_multiset Dfinsupp.toMultiset
+
+@[simp]
+theorem toMultiset_single (a : α) (n : ℕ) :
+    toMultiset (Dfinsupp.single a n) = Multiset.replicate n a :=
+  Dfinsupp.sumAddHom_single _ _ _
+#align dfinsupp.to_multiset_single Dfinsupp.toMultiset_single
+
+end Dfinsupp
+
+namespace Multiset
+
+variable [DecidableEq α]
+
+/-- A computable version of `Multiset.toFinsupp`. -/
+def toDfinsupp : Multiset α →+ Π₀ _a : α, ℕ where
+  toFun s :=
+    { toFun := fun n ↦ s.count n
+      support' := Trunc.mk ⟨s, fun i ↦ (em (i ∈ s)).imp_right Multiset.count_eq_zero_of_not_mem⟩ }
+  map_zero' := rfl
+  map_add' _ _ := Dfinsupp.ext fun _ ↦ Multiset.count_add _ _ _
+#align multiset.to_dfinsupp Multiset.toDfinsupp
+
+@[simp]
+theorem toDfinsupp_apply (s : Multiset α) (a : α) : Multiset.toDfinsupp s a = s.count a :=
+  rfl
+#align multiset.to_dfinsupp_apply Multiset.toDfinsupp_apply
+
+@[simp]
+theorem toDfinsupp_support (s : Multiset α) : s.toDfinsupp.support = s.toFinset :=
+  (Finset.filter_eq_self _).mpr fun _ hx ↦ count_ne_zero.mpr <| Multiset.mem_toFinset.1 hx
+#align multiset.to_dfinsupp_support Multiset.toDfinsupp_support
+
+@[simp]
+theorem toDfinsupp_replicate (a : α) (n : ℕ) :
+    toDfinsupp (Multiset.replicate n a) = Dfinsupp.single a n := by
+  ext i
+  dsimp [toDfinsupp]
+  simp [count_replicate, eq_comm]
+#align multiset.to_dfinsupp_replicate Multiset.toDfinsupp_replicate
+
+@[simp]
+theorem toDfinsupp_singleton (a : α) : toDfinsupp {a} = Dfinsupp.single a 1 := by
+  rw [← replicate_one, toDfinsupp_replicate]
+#align multiset.to_dfinsupp_singleton Multiset.toDfinsupp_singleton
+
+/-- `Multiset.toDfinsupp` as an `AddEquiv`. -/
+@[simps! apply symmApply]
+def equivDfinsupp : Multiset α ≃+ Π₀ _a : α, ℕ :=
+  AddMonoidHom.toAddEquiv Multiset.toDfinsupp Dfinsupp.toMultiset
+    (by
+      -- Porting note: used to be ext.
+      /- potential bug: in lean 3, `trace.ext` outputs
+         "matched goal to rule: dfinsupp.add_hom_ext'"
+         but that lemma does not apply!-/
+      apply Multiset.addHom_ext
+      simp)
+    (by
+      -- Porting note: used to be ext
+      apply Dfinsupp.add_hom_ext
+      simp)
```

Comment:
```suggestion
      apply Dfinsupp.addHom_ext
      simp)
```
(Here of course I mean the name should be fixed wherever it is defined.)

Reply:
done